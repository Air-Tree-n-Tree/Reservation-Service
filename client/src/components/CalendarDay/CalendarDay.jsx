import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import setCheckInDay from '../../store/actions/setCheckInDay.action';
import setCheckOutDay from '../../store/actions/setCheckOutDay.action';

import classes from './CalendarDay.module.css';

export const CalendarDay = ({
  status,
  dayOfMonth,
  day,
  dispatchSetCheckInDay,
  dispatchSetCheckOutDay,
  selecting,
}) => (
  <div className={[classes[status], classes.calendarDay].join(' ')}>
    <button
      className={classes.calendarDayButton}
      type="button"
      onClick={() => {
        if (selecting === 'checkin' && status === 'available') {
          dispatchSetCheckInDay(day);
        } else if (selecting === 'checkout' && status === 'available') {
          dispatchSetCheckOutDay(day);
        }
      }}
    >
      {dayOfMonth + 1}
    </button>
  </div>
);

CalendarDay.propTypes = {
  status: PropTypes.string.isRequired,
  dayOfMonth: PropTypes.number.isRequired,
  day: PropTypes.number.isRequired,
  dispatchSetCheckInDay: PropTypes.func.isRequired,
  dispatchSetCheckOutDay: PropTypes.func.isRequired,
  selecting: PropTypes.string.isRequired,
};

export const mapStateToProps = ({
  checkInDay,
  checkOutDay,
  currentDay,
  startingDay,
  availability,
  reservedDays,
  selecting,
}, { status, day }) => {
  let realStatus = status;
  const { maxNights, minNights } = availability;
  if (day < currentDay) {
    realStatus = 'unavailable';
  } else if (day === checkOutDay) {
    realStatus = 'checkoutDay';
  } else if (day === checkInDay) {
    realStatus = 'checkinDay';
  } else if (selecting === 'checkin' && realStatus === 'available') {
    for (let i = 1; i <= minNights; i += 1) {
      if (reservedDays[day - startingDay + i] === 'unavailable') {
        realStatus = 'checkoutOnly';
        break;
      }
    }
  } else if (selecting === 'checkout') {
    if (day > checkInDay + maxNights || day < checkInDay) {
      realStatus = 'unavailable';
    }
  }
  if (day > checkInDay && day < checkOutDay) {
    realStatus = 'selected';
  }
  return {
    status: realStatus,
    selecting,
  };
};

const mapDispatchToProps = {
  dispatchSetCheckInDay: setCheckInDay,
  dispatchSetCheckOutDay: setCheckOutDay,
};

export default connect(mapStateToProps, mapDispatchToProps)(CalendarDay);
