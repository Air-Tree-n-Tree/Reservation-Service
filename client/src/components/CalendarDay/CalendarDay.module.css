.calendarDay {
  height: 100%;
  width: 100%;
  line-height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.calendarDayButton {
  height: 100%;
  width: 100%;
  font-family: inherit;
  border: 0;
  padding: 0;
  background-color: transparent;
  color: inherit;
}

.calendarDayButton::-moz-focus-inner {
  border: 0;
}

.unavailable {
  color: rgb(185, 185, 185);
  text-decoration: line-through;
}

.unavailable button {
  cursor: auto;
}

.available, .checkoutOnly {
  color: rgb(46, 46, 46);
}

.available:hover {
  border-radius: 50%;
  border: 2px solid rgb(46, 46, 46);
}

.checkoutOnly:hover {
  border-radius: 50%;
  border: 2px solid rgb(185, 185, 185);
}

.checkinDate, .checkoutDate {
  border-radius: 50%;
  color: white;
  background-color: black;
}

.selected {
  background-color: lightgray;
}