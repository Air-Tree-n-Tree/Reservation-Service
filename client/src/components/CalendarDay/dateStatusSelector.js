import moment from 'moment';
import getLastPossibleCheckoutDate from '../../store/selectors/getLastPossibleCheckoutDate';

const dateIsReserved = (date, reservedDates) => (
  reservedDates[date] === 'unavailable' && reservedDates[date - 1] === 'unavailable'
);

const dateHasPassed = (date) => (
  date < moment().diff(moment('2000-01'), 'days')
);

const dateIsCheckoutOnly = (date, reservedDates, minNights) => {
  for (let i = 1; i <= minNights; i += 1) {
    if (reservedDates[date + i] === 'unavailable') {
      return true;
    }
  }
  return false;
};

const dateDoesNotMeetMinNights = (date, checkinDate, minNights) => (
  date > checkinDate && date - checkinDate < minNights
);

const dateExceedsLastPossibleCheckoutDate = (date, dates) => {
  const lastPossibleCheckoutDate = getLastPossibleCheckoutDate(dates);
  return lastPossibleCheckoutDate && date > lastPossibleCheckoutDate;
};

const dateStatusSelector = (dates, date) => {
  const { reservedDates, constraints, selection } = dates;
  if (dateHasPassed(date) || dateIsReserved(date, reservedDates)) {
    return 'unavailable';
  }

  const { checkoutDate, checkinDate } = selection;
  if (date === checkoutDate) {
    return 'checkoutDate';
  }
  if (date === checkinDate) {
    return 'checkinDate';
  }
  if (checkinDate < date && date < checkoutDate) {
    return 'selected';
  }

  const { selecting } = selection;
  const { minNights } = constraints;
  if (selecting === 'checkin' && dateIsCheckoutOnly(date, reservedDates, minNights)) {
    return 'checkoutOnly';
  }

  if (selecting === 'checkout') {
    if (date < checkinDate) {
      return 'unavailable';
    }
    if (dateDoesNotMeetMinNights(date, checkinDate, minNights)) {
      return 'minNights';
    }
    if (dateExceedsLastPossibleCheckoutDate(date, dates)) {
      return 'unavailable';
    }
  }

  return 'available';
};

export default dateStatusSelector;
