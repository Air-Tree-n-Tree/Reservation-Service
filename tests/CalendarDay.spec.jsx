import React from 'react';
import { shallow } from 'enzyme';
import { CalendarDay, mapStateToProps } from '../client/src/components/CalendarDay/CalendarDay';

describe('Calendar Day component', () => {
  let mockCheckInAction;
  let mockCheckOutAction;
  let calendarDay;

  beforeEach(() => {
    mockCheckInAction = jest.fn();
    mockCheckOutAction = jest.fn();
    calendarDay = shallow(
      <CalendarDay
        status="available"
        dayOfMonth={0}
        day={7000}
        dispatchSetCheckInDay={mockCheckInAction}
        dispatchSetCheckOutDay={mockCheckOutAction}
        selecting="checkin"
      />,
    );
  });

  it('should render correctly', () => {
    expect(calendarDay).toMatchSnapshot();
  });
  it('should dispatch actions on click', () => {
    calendarDay.find('button').simulate('click');
    expect(mockCheckInAction).toHaveBeenCalled();
    calendarDay.setProps({ selecting: 'checkout' });
    calendarDay.find('button').simulate('click');
    expect(mockCheckOutAction).toHaveBeenCalled();
  });
});

describe('Calendar Day component state/props', () => {
  it('should correctly map state to props', () => {
    const result = mapStateToProps({
      // Redux state
      checkInDay: null,
      checkOutDay: null,
      currentDay: 1,
      startingDay: 1,
      availability: {},
      reservedDays: [],
      selecting: 'checkin',
    },
    {
      // Own props
      status: 'available',
      day: 1,
    });
    expect(result).toStrictEqual(expect.objectContaining({
      status: expect.any(String),
      selecting: expect.any(String),
    }));
  });

  it('should deduce selected date', () => {
    const result = mapStateToProps({
      // Redux state
      checkInDay: 1,
      checkOutDay: 3,
      currentDay: 1,
      startingDay: 1,
      availability: {},
      reservedDays: [],
      selecting: 'checkin',
    },
    {
      // Own props
      status: 'available',
      day: 2,
    });
    expect(result.status).toBe('selected');
  });

  it('should deduce unavailable date', () => {
    const result = mapStateToProps({
      // Redux state
      checkInDay: null,
      checkOutDay: null,
      currentDay: 1,
      startingDay: 1,
      availability: {},
      reservedDays: [],
      selecting: 'checkin',
    },
    {
      // Own props
      status: 'available',
      day: 0,
    });
    expect(result.status).toBe('unavailable');
  });

  it('should deduce checkin date', () => {
    const result = mapStateToProps({
      // Redux state
      checkInDay: 1,
      checkOutDay: 3,
      currentDay: 1,
      startingDay: 1,
      availability: {},
      reservedDays: [],
      selecting: 'checkin',
    },
    {
      // Own props
      status: 'available',
      day: 1,
    });
    expect(result.status).toBe('checkinDay');
  });

  it('should deduce checkout date', () => {
    const result = mapStateToProps({
      // Redux state
      checkInDay: 1,
      checkOutDay: 3,
      currentDay: 1,
      startingDay: 1,
      availability: {},
      reservedDays: [],
      selecting: 'checkin',
    },
    {
      // Own props
      status: 'available',
      day: 3,
    });
    expect(result.status).toBe('checkoutDay');
  });

  it('should deduce checkoutOnly', () => {
    const result = mapStateToProps({
      // Redux state
      checkInDay: null,
      checkOutDay: null,
      currentDay: 0,
      startingDay: 0,
      availability: { minNights: 3 },
      reservedDays: ['available', 'unavailable'],
      selecting: 'checkin',
    },
    {
      // Own props
      status: 'available',
      day: 0,
    });
    expect(result.status).toBe('checkoutOnly');
  });
});
