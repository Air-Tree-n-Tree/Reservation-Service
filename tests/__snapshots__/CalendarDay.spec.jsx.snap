// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Day component should render correctly 1`] = `
<div
  className="available calendarDay"
>
  <button
    className="calendarDayButton"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
</div>
`;
